keel-core/src/main/kotlin/com/netflix/spinnaker/keel/artifacts/BranchAndTimestampSortingStrategy.kt

```diff
@@ -0,0 +1,21 @@
+package com.netflix.spinnaker.keel.artifacts
+
+import com.netflix.spinnaker.keel.api.artifacts.PublishedArtifact
+import com.netflix.spinnaker.keel.api.artifacts.SortingStrategy
+
+/**
+ * A [SortingStrategy] that compares artifact versions by branch and timestamp.
+ */
+object BranchAndTimestampSortingStrategy : SortingStrategy {
+  override val type: String = "branch-and-timestamp"
+
+  override val comparator: Comparator<PublishedArtifact> =
+    compareBy<PublishedArtifact> { it.gitMetadata?.branch }.thenByDescending { it.createdAt }
```

Comment:
This is it... All this work to prepare for this single line right here... 🙂

Reply:
@erikmunson A question for you. Given this set of versions, branches and timestamps (in order of timestamp):

|version|branch|createdAt|
--|--|--
1.0.0|master|2020-11-12T15:20:00Z
1.0.1-rc.1|mybranch|2020-11-12T16:20:00Z
1.0.1|master|2020-11-12T17:20:00Z
1.0.2-rc.1|mybranch|2020-11-12T18:20:00Z

...this comparator will return the following list:

|version|branch|createdAt|
--|--|--
1.0.0|master|2020-11-12T15:20:00Z
1.0.1|master|2020-11-12T17:20:00Z
1.0.1-rc.1|mybranch|2020-11-12T16:20:00Z
1.0.2-rc.1|mybranch|2020-11-12T18:20:00Z

That is, we group by branch, _then_ sort by timestamp. When getting artifact summaries for the UI, given a size limit for the list, we may end up returning a bunch of versions for the first branch, and few or none for the second, in this example. From a UX perspective, I'm not sure this is ideal. We could ignore the branch altogether and sort only by timestamp. We could also roll this out and see what it looks like in staging, and then decide. Thoughts?

Reply:
P.S. To be clear, this is a compound problem between the sorting comparator here and the implementation of that API, so it's a bit tangential to this PR, but I was reminded of it when reviewing my own code, and thought I'd bring it up.

Reply:
Is there any sort of mechanical/functional usefulness on the keel side in sorting by branch here? As a human using this I would expect things to be used strictly chronologically — i.e. if i have that `regex: ".*-branch"` filter, it would be pretty strange if I pushed a matching version and it got stuck in the middle of the list behind other things from days ago simply because of how the branch name happened to sort alphabetically against the older branch names. If there _is_ some kind of functional reason this branch sorting is useful for keel, we can talk about the tradeoffs/other alternatives.

Reply:
In hindsight, I don't think there is. I think at some point I mixed up filtering by branch/PR with sorting by timestamp in my head, and it stuck. Let's agree to change it to `CreatedAtSortingStrategy` and sort by timestamp only, then?

Reply:
Love it! ✅